// @injectivelabs/x402/dist/networks/index.js
/**
 * Injective EVM network definitions for the x402 protocol.
 *
 * Chain IDs:
 *   Mainnet: 1776  (CAIP-2: eip155:1776)
 *   Testnet: 1439  (CAIP-2: eip155:1439)
 */
// ─── Chain definitions (viem-compatible) ─────────────────────────────────────
export const injectiveEvm = {
    id: 1776,
    name: "Injective EVM",
    nativeCurrency: { name: "Injective", symbol: "INJ", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://sentry.evm-rpc.injective.network"] },
        public: { http: ["https://sentry.evm-rpc.injective.network"] },
    },
    blockExplorers: {
        default: { name: "Blockscout", url: "https://blockscout.injective.network" },
    },
    contracts: {},
};
export const injectiveEvmTestnet = {
    id: 1439,
    name: "Injective EVM Testnet",
    nativeCurrency: { name: "Injective", symbol: "INJ", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://k8s.testnet.json-rpc.injective.network"] },
        public: { http: ["https://k8s.testnet.json-rpc.injective.network"] },
    },
    blockExplorers: {
        default: { name: "Blockscout Testnet", url: "https://testnet.blockscout.injective.network" },
    },
    testnet: true,
    contracts: {},
};
// ─── CAIP-2 identifiers ───────────────────────────────────────────────────────
export const INJECTIVE_MAINNET_CAIP2 = "eip155:1776";
export const INJECTIVE_TESTNET_CAIP2 = "eip155:1439";
/**
 * Known token addresses on Injective EVM.
 *
 * Status of EIP-3009 (transferWithAuthorization) on native tokens:
 *   - USDC mainnet (0xa00C…235a): Circle FiatTokenV2_2 — full EIP-3009 + EIP-2612.
 *                                  Use this for x402 payments. Domain version "2".
 *   - USDC testnet (0x0C38…4C5d): Circle FiatTokenV2_2 — same shape as mainnet.
 *   - USDT (0x88f7…Cc13): plain IBC-bridged ERC-20 — NO EIP-3009.
 *   - WINJ (0x0000…FfB):  plain ERC-20 — NO EIP-3009.
 *
 * Non-EIP-3009 tokens such as USDT and WINJ cannot be used directly with x402;
 * they are listed for informational purposes only.
 *
 * The legacy IBC-bridged USDC at 0x2a25…3989 is intentionally not registered
 * here. It has no EIP-3009 and is superseded by the native Circle deployment.
 *
 * For any custom EIP-3009 token your integration needs, register it at startup
 * via {@link registerToken} rather than setting environment variables — that
 * keeps this module pure and safe to bundle for browsers.
 */
export const TOKENS = {
    [INJECTIVE_MAINNET_CAIP2]: {
        // Native USDC — Circle FiatTokenV2_2, full EIP-3009. name() returned "USDC" on 2026-05-12.
        USDC: {
            address: "0xa00C59fF5a080D2b954d0c75e46E22a0c371235a",
            decimals: 6,
            symbol: "USDC",
            name: "USDC",
            eip3009: true,
            eip712Version: "2",
        },
        // Plain ERC-20s (NO EIP-3009 — cannot be used directly with x402)
        USDT: {
            address: "0x88f7F2b685F9692caf8c478f5BADF09eE9B1Cc13",
            decimals: 6,
            symbol: "USDT",
            name: "Tether",
            eip3009: false,
        },
        WINJ: {
            address: "0x0000000088827d2d103ee2d9A6b781773AE03FfB",
            decimals: 18,
            symbol: "WINJ",
            name: "Wrapped INJ",
            eip3009: false,
        },
    },
    [INJECTIVE_TESTNET_CAIP2]: {
        // Native USDC on testnet — Circle FiatTokenV2_2 (probed: same shape as
        // mainnet, version "2", DOMAIN_SEPARATOR non-zero). Use directly with x402.
        USDC: {
            address: "0x0C382e685bbeeFE5d3d9C29e29E341fEE8E84C5d",
            decimals: 6,
            symbol: "USDC",
            name: "USDC",
            eip3009: true,
            eip712Version: "2",
        },
        // Custom testnet EIP-3009 tokens can be registered at app startup via
        // registerToken() — addresses are not canonical.
    },
};
/**
 * Register (or override) a token entry at runtime.
 *
 * Useful when your integration uses a custom EIP-3009 token that is not in the
 * canonical registry yet. Keeps this module pure — no `process.env` reads at
 * import time, which means it is safe to bundle for browsers or Deno/Bun
 * runtimes.
 *
 * @example
 *   registerToken("eip155:1776", "MYTOKEN", {
 *     address:       "0x…",
 *     decimals:      6,
 *     symbol:        "MYTOKEN",
 *     name:          "My EIP-3009 Token",
 *     eip3009:       true,
 *     eip712Version: "2",
 *   });
 */
export function registerToken(network, symbol, config) {
    TOKENS[network] = TOKENS[network] ?? {};
    TOKENS[network][symbol.toUpperCase()] = config;
}
// ─── Network helpers ──────────────────────────────────────────────────────────
export function isInjectiveNetwork(network) {
    return (network === INJECTIVE_MAINNET_CAIP2 ||
        network === INJECTIVE_TESTNET_CAIP2);
}
export function getViemChain(network) {
    return network === INJECTIVE_MAINNET_CAIP2 ? injectiveEvm : injectiveEvmTestnet;
}
export function getRpcUrl(network) {
    return network === INJECTIVE_MAINNET_CAIP2
        ? "https://sentry.evm-rpc.injective.network"
        : "https://k8s.testnet.json-rpc.injective.network";
}
export function getToken(network, symbol) {
    return TOKENS[network]?.[symbol.toUpperCase()];
}
export function getTokenByAddress(network, address) {
    return Object.values(TOKENS[network] ?? {}).find((t) => t.address.toLowerCase() === address.toLowerCase());
}
