// @injectivelabs/x402/dist/middleware/index.js
/**
 * Express middleware for x402 payments on Injective EVM.
 *
 * Usage:
 *   app.use(injectivePaymentMiddleware(routes, facilitatorConfig));
 *
 * On a protected route:
 *   - Requests without a `PAYMENT-SIGNATURE` header receive a 402 whose JSON
 *     body lists accepted PaymentRequirements (the `accepts` field) and whose
 *     `PAYMENT-REQUIRED` header carries the same body encoded as base64.
 *   - Requests with a valid `PAYMENT-SIGNATURE` header are verified + settled,
 *     then the handler runs. The response carries a `PAYMENT-RESPONSE` header
 *     with a base64-encoded settlement receipt
 *     ({success, transaction, network, payer}).
 *   - Invalid payments get a 402 with the rejection reason.
 *
 * For temporary compatibility, `X-PAYMENT` is still accepted on input and
 * `X-PAYMENT-RESPONSE` is mirrored on output. New integrations should use the
 * upstream v2 `PAYMENT-*` header names.
 */
import { privateKeyToAccount } from "viem/accounts";
import { InjectiveFacilitator } from "../facilitator/index.js";
import { decodePaymentSignatureHeader } from "../client/index.js";
import { createFacilitatorRequest } from "../protocol/index.js";
import { getTokenByAddress, isInjectiveNetwork, } from "../networks/index.js";
import { SettleResponseSchema, VerifyResponseSchema, } from "../schemas.js";
/**
 * Express middleware factory for x402 payment protection.
 */
export function injectivePaymentMiddleware(routes, options) {
    // Validate options
    if (!options.facilitatorUrl && !options.facilitator) {
        throw new Error("injectivePaymentMiddleware: provide either facilitatorUrl or facilitator config");
    }
    const localFacilitator = options.facilitator
        ? new InjectiveFacilitator(options.facilitator)
        : null;
    // Derive facilitator address from private key so it can be the default payTo
    const facilitatorAddress = options.facilitator?.privateKey
        ? privateKeyToAccount(options.facilitator.privateKey).address
        : undefined;
    validatePayToDefaults(routes, facilitatorAddress);
    return async (req, res, next) => {
        const routeKey = `${req.method} ${req.path}`;
        const config = routes[routeKey] ?? matchWildcard(routes, req.method, req.path);
        // Route not protected — pass through
        if (!config) {
            return next();
        }
        const resource = buildResourceUrl(req, options.baseUrl);
        const requirements = buildRequirements(config, resource, facilitatorAddress);
        // Prefer the v2 PAYMENT-SIGNATURE header while temporarily accepting X-PAYMENT.
        const paymentHeader = req.headers["payment-signature"] ??
            req.headers["x-payment"];
        // No payment — return 402
        if (!paymentHeader) {
            return send402(res, requirements, resource, config);
        }
        // Decode and verify payment
        let paymentPayload;
        try {
            paymentPayload = decodePaymentSignatureHeader(paymentHeader);
        }
        catch {
            return send402(res, requirements, resource, config, "Invalid PAYMENT-SIGNATURE header encoding");
        }
        // Try every requirement paying this recipient. Network/asset are bound by
        // the EIP-712 signature domain, so verification is the source of truth for
        // the exact offered requirement a client chose.
        const candidateReqs = requirements.filter((r) => r.payTo.toLowerCase() === paymentPayload.payload.authorization.to.toLowerCase());
        if (candidateReqs.length === 0) {
            return send402(res, requirements, resource, config, "Payment network not accepted");
        }
        let matchedReq;
        const invalidReasons = [];
        for (const candidateReq of candidateReqs) {
            const verifyResult = await verify({ paymentPayload, paymentRequirements: candidateReq }, localFacilitator, options.facilitatorUrl);
            if (verifyResult.isValid) {
                matchedReq = candidateReq;
                break;
            }
            if (verifyResult.invalidReason) {
                invalidReasons.push(`${candidateReq.network}/${candidateReq.asset}: ${verifyResult.invalidReason}`);
            }
        }
        if (!matchedReq) {
            return send402(res, requirements, resource, config, invalidReasons[0] ?? "Payment requirements not accepted");
        }
        const policy = options.settlementPolicy ?? "before";
        if (policy === "before") {
            // Pay first, then let the handler run.
            const settleResult = await settle({ paymentPayload, paymentRequirements: matchedReq }, localFacilitator, options.facilitatorUrl);
            if (!settleResult.success) {
                setReceiptHeader(res, settleResult);
                res.status(402).json({
                    error: "payment_settlement_failed",
                    message: settleResult.errorMessage ?? settleResult.errorReason,
                    settlement: settleResult,
                });
                return;
            }
            setReceiptHeader(res, settleResult);
            attachPaymentMeta(req, matchedReq, paymentPayload, settleResult.transaction || undefined);
            return next();
        }
        // policy === "after-success"
        // Run the handler first; only settle if the response status is 2xx.
        // Buffer writes so no paid content reaches the client before settlement.
        attachPaymentMeta(req, matchedReq, paymentPayload, undefined);
        const bufferedChunks = [];
        const originalWrite = res.write.bind(res);
        const originalEnd = res.end.bind(res);
        const originalWriteHead = res.writeHead.bind(res);
        const originalFlushHeaders = res.flushHeaders.bind(res);
        let recordedWriteHead;
        let ending = false;
        res.write = function writeInterceptor(...writeArgs) {
            const callback = extractCallback(writeArgs);
            const chunk = writeArgs[0];
            const encoding = extractEncoding(writeArgs);
            if (chunk !== undefined && typeof chunk !== "function") {
                bufferedChunks.push({ chunk, encoding });
            }
            if (callback)
                queueMicrotask(callback);
            return true;
        };
        res.flushHeaders = function flushHeadersInterceptor() {
            // Delay header emission until settlement decides whether to release or replace the body.
        };
        res.writeHead = function writeHeadInterceptor(...writeHeadArgs) {
            recordedWriteHead = { args: writeHeadArgs };
            const statusCode = writeHeadArgs[0];
            const statusMessage = writeHeadArgs[1];
            if (typeof statusCode === "number") {
                res.statusCode = statusCode;
            }
            if (typeof statusMessage === "string") {
                res.statusMessage = statusMessage;
            }
            return res;
        };
        res.end = function endInterceptor(...endArgs) {
            if (ending)
                return originalEnd(...endArgs);
            ending = true;
            const callback = extractCallback(endArgs);
            const chunk = endArgs[0];
            const encoding = extractEncoding(endArgs);
            if (chunk !== undefined && typeof chunk !== "function") {
                bufferedChunks.push({ chunk, encoding });
            }
            const status = res.statusCode;
            if (status < 200 || status >= 300) {
                // Handler errored or redirected — skip settlement, free ride for this request.
                replayBufferedResponse(bufferedChunks, originalWrite, originalEnd, originalWriteHead, recordedWriteHead, callback);
                return res;
            }
            (async () => {
                try {
                    const settleResult = await settle({ paymentPayload, paymentRequirements: matchedReq }, localFacilitator, options.facilitatorUrl);
                    if (settleResult.success) {
                        setReceiptHeader(res, settleResult);
                        req.x402 = {
                            ...req.x402,
                            txHash: settleResult.transaction,
                        };
                        if (recordedWriteHead) {
                            replayBufferedResponse(bufferedChunks, originalWrite, originalEnd, originalWriteHead, recordedWriteHead, callback);
                        }
                        else {
                            originalFlushHeaders();
                            replayBufferedResponse(bufferedChunks, originalWrite, originalEnd, originalWriteHead, undefined, callback);
                        }
                    }
                    else {
                        // Settlement failed after a successful handler. Swap the body for
                        // a 402 — the client didn't pay, so they don't get the result.
                        sendSettlementFailure(res, settleResult, originalEnd, callback);
                    }
                }
                catch (e) {
                    const settleResult = failedSettlementResponse(e.message, matchedReq.network, paymentPayload.payload.authorization.from);
                    sendSettlementFailure(res, settleResult, originalEnd, callback);
                }
            })();
            return res;
        };
        return next();
    };
}
function setReceiptHeader(res, settleResult) {
    const encoded = Buffer.from(JSON.stringify(settleResult)).toString("base64");
    res.setHeader("PAYMENT-RESPONSE", encoded);
    res.setHeader("X-PAYMENT-RESPONSE", encoded);
}
function sendSettlementFailure(res, settleResult, originalEnd, callback) {
    setReceiptHeader(res, settleResult);
    res.statusCode = 402;
    res.removeHeader("Content-Length");
    res.removeHeader("ETag");
    res.removeHeader("Content-Encoding");
    res.removeHeader("Content-Range");
    res.removeHeader("Accept-Ranges");
    res.removeHeader("Last-Modified");
    res.setHeader("Content-Type", "application/json");
    const errBody = JSON.stringify({
        error: "payment_settlement_failed",
        message: settleResult.errorMessage ?? settleResult.errorReason,
        settlement: settleResult,
    });
    res.setHeader("Content-Length", Buffer.byteLength(errBody));
    if (callback) {
        originalEnd(errBody, callback);
        return;
    }
    originalEnd(errBody);
}
function replayBufferedResponse(chunks, originalWrite, originalEnd, originalWriteHead, recordedWriteHead, callback) {
    if (recordedWriteHead) {
        originalWriteHead(...recordedWriteHead.args);
    }
    for (const { chunk, encoding } of chunks) {
        if (encoding) {
            originalWrite(chunk, encoding);
        }
        else {
            originalWrite(chunk);
        }
    }
    if (callback) {
        originalEnd(callback);
        return;
    }
    originalEnd();
}
function extractCallback(args) {
    const last = args[args.length - 1];
    return typeof last === "function" ? last : undefined;
}
function extractEncoding(args) {
    return typeof args[1] === "string" ? args[1] : undefined;
}
function attachPaymentMeta(req, matchedReq, paymentPayload, txHash) {
    req.x402 = {
        payer: paymentPayload.payload.authorization.from,
        network: matchedReq.network,
        amount: matchedReq.amount,
        asset: matchedReq.asset,
        txHash,
    };
}
// ─── Helpers ──────────────────────────────────────────────────────────────────
function buildRequirements(config, resource, facilitatorAddress) {
    return config.accepts.map((opt) => {
        const token = isInjectiveNetwork(opt.network)
            ? getTokenByAddress(opt.network, opt.asset)
            : undefined;
        return {
            scheme: "exact",
            network: opt.network,
            amount: opt.amount,
            payTo: opt.payTo ?? facilitatorAddress,
            maxTimeoutSeconds: opt.maxTimeoutSeconds ?? 60,
            asset: opt.asset,
            extra: {
                name: token?.name ?? "USDC",
                version: token?.eip712Version ?? "2",
                assetTransferMethod: "eip3009",
            },
        };
    });
}
function validatePayToDefaults(routes, facilitatorAddress) {
    if (facilitatorAddress)
        return;
    for (const [routeKey, config] of Object.entries(routes)) {
        for (const opt of config.accepts) {
            if (!opt.payTo) {
                throw new Error(`injectivePaymentMiddleware: route ${routeKey} must set payTo ` +
                    "when using facilitatorUrl without a local facilitator private key");
            }
        }
    }
}
function send402(res, requirements, resource, config, reason) {
    const body = {
        x402Version: 2,
        error: reason ?? "PAYMENT-SIGNATURE header is required",
        resource: {
            url: resource,
            description: config.description,
            mimeType: config.mimeType ?? "application/json",
        },
        accepts: requirements,
    };
    const encoded = Buffer.from(JSON.stringify(body)).toString("base64");
    res.setHeader("PAYMENT-REQUIRED", encoded);
    res.status(402).json(body);
}
function buildResourceUrl(req, baseUrl) {
    if (baseUrl)
        return `${baseUrl}${req.path}`;
    const protocol = req.protocol;
    const host = req.get("host") ?? "localhost";
    return `${protocol}://${host}${req.path}`;
}
/**
 * Route pattern syntax supported:
 *   - Literal paths          e.g. "/api/weather"
 *   - Named parameters       e.g. "/api/users/:id"     (matches any single segment)
 *   - Trailing catch-all     e.g. "/api/ai/*"          (matches any suffix)
 *
 * Patterns are compiled to regexes on first use and cached for the lifetime
 * of the middleware instance.
 */
const patternCache = new WeakMap();
function compilePattern(pattern) {
    // Escape regex metacharacters except the ones we handle ourselves.
    const escaped = pattern.replace(/[.+?^${}()|[\]\\]/g, "\\$&");
    // `/:param` → `/([^/]+)` ; trailing `*` → `(.*)`
    const withParams = escaped
        .replace(/:([A-Za-z_][A-Za-z0-9_]*)/g, "([^/]+)")
        .replace(/\\\*$/, ".*");
    return new RegExp(`^${withParams}$`);
}
function matchWildcard(routes, method, path) {
    let cache = patternCache.get(routes);
    if (!cache) {
        cache = new Map();
        patternCache.set(routes, cache);
    }
    for (const [key, config] of Object.entries(routes)) {
        const spaceIdx = key.indexOf(" ");
        if (spaceIdx < 0)
            continue;
        const routeMethod = key.slice(0, spaceIdx);
        const routePath = key.slice(spaceIdx + 1);
        if (routeMethod !== method)
            continue;
        let re = cache.get(key);
        if (!re) {
            re = compilePattern(routePath);
            cache.set(key, re);
        }
        if (re.test(path))
            return config;
    }
    return undefined;
}
async function verify(req, local, facilitatorUrl) {
    if (local) {
        return local.verify(req);
    }
    if (!facilitatorUrl) {
        return invalidVerifyResponse("Remote facilitator URL is not configured");
    }
    try {
        const resp = await globalThis.fetch(`${facilitatorUrl}/verify`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(createFacilitatorRequest(req)),
        });
        if (!resp.ok) {
            return invalidVerifyResponse(`Remote facilitator verify failed with HTTP ${resp.status}`);
        }
        const parsed = VerifyResponseSchema.safeParse(await resp.json());
        if (!parsed.success) {
            return invalidVerifyResponse(`Remote facilitator returned invalid verify response: ${parsed.error.issues[0]?.message ?? "schema validation failed"}`);
        }
        return parsed.data;
    }
    catch (e) {
        return invalidVerifyResponse(e.message);
    }
}
async function settle(req, local, facilitatorUrl) {
    if (local) {
        return local.settle(req);
    }
    const expectedNetwork = req.paymentRequirements.network;
    const expectedPayer = req.paymentPayload.payload.authorization.from;
    if (!facilitatorUrl) {
        return failedSettlementResponse("Remote facilitator URL is not configured", expectedNetwork, expectedPayer);
    }
    try {
        const resp = await globalThis.fetch(`${facilitatorUrl}/settle`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(createFacilitatorRequest(req)),
        });
        if (!resp.ok) {
            return failedSettlementResponse(`Remote facilitator settle failed with HTTP ${resp.status}`, expectedNetwork, expectedPayer);
        }
        const parsed = SettleResponseSchema.safeParse(await resp.json());
        if (!parsed.success) {
            return failedSettlementResponse(`Remote facilitator returned invalid settle response: ${parsed.error.issues[0]?.message ?? "schema validation failed"}`, expectedNetwork, expectedPayer);
        }
        const settleResult = parsed.data;
        if (settleResult.success) {
            const payerMatches = settleResult.payer?.toLowerCase() === expectedPayer.toLowerCase();
            if (!settleResult.transaction) {
                return failedSettlementResponse("Remote facilitator settlement response missing transaction", expectedNetwork, expectedPayer);
            }
            if (settleResult.network !== expectedNetwork) {
                return failedSettlementResponse("Remote facilitator settlement response network mismatch", expectedNetwork, expectedPayer);
            }
            if (!payerMatches) {
                return failedSettlementResponse("Remote facilitator settlement response payer mismatch", expectedNetwork, expectedPayer);
            }
        }
        return settleResult;
    }
    catch (e) {
        return failedSettlementResponse(e.message, expectedNetwork, expectedPayer);
    }
}
function invalidVerifyResponse(message) {
    return {
        isValid: false,
        invalidReason: "remote_facilitator_error",
        invalidMessage: message,
    };
}
function failedSettlementResponse(message, network, payer) {
    return {
        success: false,
        errorReason: "remote_facilitator_error",
        errorMessage: message,
        payer,
        transaction: "",
        network,
    };
}
