// @injectivelabs/x402/dist/client/index.js
/**
 * x402 Client for Injective EVM.
 *
 * Handles the client side of the payment flow:
 *   1. Receive a 402 response with payment requirements
 *   2. Create and sign an EIP-3009 authorization
 *   3. Retry the original request with the PAYMENT header
 *
 * Usage:
 *   const client = createInjectiveClient({ privateKey: "0x..." });
 *   const response = await client.fetch("https://api.example.com/paid-endpoint");
 */
import { createPublicClient, createWalletClient, http, } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { getTokenByAddress, getRpcUrl, getViemChain, isInjectiveNetwork, } from "../networks/index.js";
import { signAuthorization, createNonce, getTokenName, } from "../eip3009.js";
import { PaymentPayloadSchema, PaymentRequiredSchema, assertPayloadSize, } from "../schemas.js";
// ─── Payment creation ─────────────────────────────────────────────────────────
/**
 * Build and sign an x402 payment payload for the given requirements.
 *
 * @param config       Client configuration (private key, optional RPC override)
 * @param requirements Payment requirements from a 402 response
 * @returns            Signed payload ready to base64-encode into the PAYMENT header
 */
export async function createPayment(config, requirements) {
    const network = requirements.network;
    if (!isInjectiveNetwork(network)) {
        throw new Error(`createPayment: unsupported network ${network}`);
    }
    const chain = getViemChain(network);
    const rpcUrl = config.rpcUrl ?? getRpcUrl(network);
    const account = privateKeyToAccount(config.privateKey);
    const publicClient = createPublicClient({ chain, transport: http(rpcUrl) });
    const walletClient = createWalletClient({ account, chain, transport: http(rpcUrl) });
    const tokenAddress = requirements.asset;
    const tokenVersion = resolveTokenVersion(requirements);
    const tokenName = await getTokenName(publicClient, tokenAddress);
    // Build time window: valid from ~now (minus 10s grace) for maxTimeoutSeconds
    const now = BigInt(Math.floor(Date.now() / 1000));
    const validAfter = now - 10n;
    const validBefore = now + BigInt(requirements.maxTimeoutSeconds);
    const nonce = createNonce();
    const value = BigInt(requirements.amount);
    const auth = {
        from: account.address,
        to: requirements.payTo,
        value,
        validAfter,
        validBefore,
        nonce,
    };
    const signature = await signAuthorization(walletClient, tokenAddress, tokenName, chain.id, auth, tokenVersion);
    return {
        x402Version: 2,
        accepted: requirements,
        payload: {
            signature,
            authorization: {
                from: auth.from,
                to: auth.to,
                value: auth.value.toString(),
                validAfter: auth.validAfter.toString(),
                validBefore: auth.validBefore.toString(),
                nonce: auth.nonce,
            },
        },
    };
}
/**
 * Encode a payment payload as a base64 string for the PAYMENT-SIGNATURE HTTP header.
 */
export function encodePaymentSignatureHeader(payload) {
    return Buffer.from(JSON.stringify(PaymentPayloadSchema.parse(payload))).toString("base64");
}
/**
 * Decode a base64 PAYMENT-SIGNATURE header back into a payload object.
 *
 * Validates size and shape via zod so a malformed or oversized header doesn't
 * propagate untyped garbage deeper into the facilitator. Throws on mismatch.
 */
export function decodePaymentSignatureHeader(header) {
    assertPayloadSize(header);
    const json = Buffer.from(header, "base64").toString("utf-8");
    assertPayloadSize(json);
    const raw = JSON.parse(json);
    return PaymentPayloadSchema.parse(raw);
}
/**
 * Parse an x402 PaymentRequired object from a 402 response header.
 */
export function parsePaymentRequired(header) {
    assertPayloadSize(header);
    const json = Buffer.from(header, "base64").toString("utf-8");
    assertPayloadSize(json);
    const raw = JSON.parse(json);
    return PaymentRequiredSchema.parse(raw);
}
/**
 * Create an x402-aware fetch client that automatically handles 402 responses
 * on Injective EVM networks.
 */
export function createInjectiveClient(options) {
    const preferredNetworks = options.preferredNetworks ?? [
        "eip155:1776",
        "eip155:1439",
    ];
    return {
        /**
         * Fetch a resource, automatically paying if a 402 is returned.
         */
        async fetch(url, init = {}) {
            // First attempt — no payment
            const firstResponse = await globalThis.fetch(url, init);
            if (firstResponse.status !== 402) {
                return firstResponse;
            }
            // Per the x402 spec, requirements come in the response body's `accepts`
            // field. Fall back to the legacy base64 `PAYMENT-REQUIRED` header for
            // older servers that only emit the header.
            const allRequirements = await read402Requirements(firstResponse);
            if (!allRequirements) {
                throw new Error("402 response has no recognizable payment requirements " +
                    "(missing body.accepts and PAYMENT-REQUIRED header)");
            }
            // Pick the first requirement matching our preferred networks
            const chosen = pickRequirements(allRequirements, preferredNetworks);
            if (!chosen) {
                throw new Error(`No supported Injective network in payment requirements. ` +
                    `Available: ${allRequirements.map((r) => r.network).join(", ")}`);
            }
            // Create and sign payment
            const payload = await createPayment(options, chosen);
            const paymentHeader = encodePaymentSignatureHeader(payload);
            // Retry with the v2 payment signature header while temporarily retaining
            // `X-PAYMENT` for compatibility with older servers.
            const headers = new Headers(init.headers);
            headers.set("PAYMENT-SIGNATURE", paymentHeader);
            headers.set("X-PAYMENT", paymentHeader);
            return globalThis.fetch(url, { ...init, headers });
        },
    };
}
// ─── Helpers ──────────────────────────────────────────────────────────────────
function pickRequirements(all, preferred) {
    for (const network of preferred) {
        const match = all.find((r) => r.network === network);
        if (match)
            return match;
    }
    // Fall back to any Injective network
    return all.find((r) => isInjectiveNetwork(r.network));
}
function resolveTokenVersion(requirements) {
    const network = requirements.network;
    const registryToken = getTokenByAddress(network, requirements.asset);
    const version = registryToken?.eip712Version ??
        readVersionExtra(requirements.extra);
    if (!version) {
        throw new Error(`createPayment: unknown asset ${requirements.asset} requires ` +
            "paymentRequirements.extra.version for EIP-712 signing");
    }
    return version;
}
function readVersionExtra(extra) {
    return typeof extra?.version === "string" ? extra.version : undefined;
}
/**
 * Extract payment requirements from a 402 response, trying the spec-compliant
 * body shape first and falling back to the legacy `PAYMENT-REQUIRED` header.
 * Returns `undefined` if neither source is present.
 */
async function read402Requirements(res) {
    try {
        return PaymentRequiredSchema.parse(await res.clone().json()).accepts;
    }
    catch {
        // body wasn't JSON — fall through to header
    }
    const headerValue = res.headers.get("PAYMENT-REQUIRED");
    if (!headerValue)
        return undefined;
    try {
        return parsePaymentRequired(headerValue).accepts;
    }
    catch {
        return undefined;
    }
}
export function parsePaymentResponseHeader(res) {
    const h = res.headers.get("PAYMENT-RESPONSE") ?? res.headers.get("X-PAYMENT-RESPONSE");
    if (!h)
        return undefined;
    try {
        return JSON.parse(Buffer.from(h, "base64").toString("utf-8"));
    }
    catch {
        return undefined;
    }
}
