// @injectivelabs/x402/dist/protocol/index.js
import { createPublicClient, http, } from "viem";
import { getTokenByAddress, getViemChain, isInjectiveNetwork, } from "../networks/index.js";
import { FacilitatorVerifyRequestSchema } from "../schemas.js";
import { getTokenBalance, getTokenName, isNonceUsed as readNonceUsed, verifyAuthorization, } from "../eip3009.js";
export { buildTransferWithAuthorizationTypedData, encodeTransferWithAuthorizationCalldata, } from "../eip3009.js";
export function toInnerVerifyRequest(request) {
    return {
        paymentPayload: request.paymentPayload,
        paymentRequirements: request.paymentRequirements,
    };
}
export function createPaymentPayload(accepted, payload, resource) {
    return {
        x402Version: 2,
        ...(resource ? { resource } : {}),
        accepted,
        payload,
    };
}
export function createFacilitatorRequest(inner) {
    return {
        x402Version: 2,
        paymentPayload: inner.paymentPayload,
        paymentRequirements: inner.paymentRequirements,
    };
}
export function normalizeFacilitatorRequest(body) {
    const parsed = FacilitatorVerifyRequestSchema.safeParse(body);
    if (!parsed.success) {
        const message = parsed.error.issues[0]?.message ?? "unknown error";
        throw new Error(`Invalid x402 v2 facilitator request: ${message}`);
    }
    return {
        ...parsed.data,
        inner: toInnerVerifyRequest(parsed.data),
    };
}
export async function verifyPaymentRequest(request, options = {}) {
    const { paymentPayload, paymentRequirements } = request;
    const network = paymentRequirements.network;
    const auth = paymentPayload.payload.authorization;
    const payer = auth.from;
    if (!isInjectiveNetwork(network)) {
        return invalid("unsupported_network", undefined, undefined, `Unsupported network: ${network}. Expected eip155:1776 or eip155:1439`);
    }
    const acceptedMismatch = validateAcceptedMatchesRequirements(paymentPayload.accepted, paymentRequirements);
    if (acceptedMismatch) {
        return invalid(acceptedMismatch.invalidReason, payer, undefined, acceptedMismatch.invalidMessage);
    }
    if (auth.to.toLowerCase() !== paymentRequirements.payTo.toLowerCase()) {
        return invalid("payment_recipient_mismatch", payer, undefined, "Payment recipient mismatch");
    }
    if (BigInt(auth.value) !== BigInt(paymentRequirements.amount)) {
        return invalid("payment_amount_mismatch", payer, undefined, "Payment amount mismatch");
    }
    const domainError = validateDomainExtra(paymentRequirements.extra, paymentRequirements.asset, network);
    if (domainError) {
        return invalid("payment_eip712_domain_mismatch", payer, undefined, domainError);
    }
    const allowed = options.allowedAssets?.map((asset) => asset.toLowerCase());
    if (allowed && allowed.length > 0 && !allowed.includes(paymentRequirements.asset.toLowerCase())) {
        return invalid("asset_not_allowed", payer, undefined, `Asset not allowed by facilitator: ${paymentRequirements.asset}`);
    }
    const minimum = options.minPaymentPerAsset?.[paymentRequirements.asset.toLowerCase()];
    if (minimum !== undefined && BigInt(paymentRequirements.amount) < BigInt(minimum)) {
        return invalid("payment_below_minimum", payer, undefined, `Payment below facilitator minimum: ${paymentRequirements.amount} < ${minimum}`);
    }
    const registryToken = getTokenByAddress(network, paymentRequirements.asset);
    const tokenVersion = registryToken?.eip712Version ?? readVersionExtra(paymentRequirements.extra);
    if (!tokenVersion) {
        return invalid("payment_eip712_domain_mismatch", payer, undefined, `Unknown asset ${paymentRequirements.asset} requires explicit EIP-712 version in paymentRequirements.extra.version`);
    }
    const now = BigInt(options.now?.() ?? Math.floor(Date.now() / 1000));
    if (BigInt(auth.validAfter) > now) {
        return invalid("payment_not_yet_valid", payer, undefined, "Payment authorization is not yet valid");
    }
    if (BigInt(auth.validBefore) < now) {
        return invalid("payment_expired", payer, undefined, "Payment authorization expired");
    }
    let nonceUsed;
    try {
        nonceUsed = await (options.isNonceUsed ?? defaultNonceReader)(paymentRequirements.asset, payer, auth.nonce, network);
    }
    catch {
        return invalid("nonce_check_failed", payer, undefined, "Unable to verify payment nonce");
    }
    if (nonceUsed) {
        return invalid("nonce_already_used", payer, undefined, "Nonce has already been used");
    }
    let balance;
    try {
        balance = await (options.getBalance ?? defaultBalanceReader)(paymentRequirements.asset, payer, network);
    }
    catch {
        return invalid("balance_check_failed", payer, undefined, "Unable to verify payer balance");
    }
    if (balance < BigInt(paymentRequirements.amount)) {
        return invalid("insufficient_funds", payer, balance, "Insufficient payer balance");
    }
    let tokenName;
    try {
        tokenName =
            registryToken?.name ??
                await (options.getTokenName ?? defaultTokenNameReader)(paymentRequirements.asset, network);
    }
    catch {
        return invalid("token_name_check_failed", payer, balance, "Unable to verify token name");
    }
    let signer;
    try {
        signer = await verifyAuthorization(paymentRequirements.asset, tokenName, chainIdFromNetwork(network), {
            from: payer,
            to: auth.to,
            value: BigInt(auth.value),
            validAfter: BigInt(auth.validAfter),
            validBefore: BigInt(auth.validBefore),
            nonce: auth.nonce,
        }, paymentPayload.payload.signature, tokenVersion);
    }
    catch {
        return invalid("signature_verification_failed", payer, balance, "Payment signature verification failed");
    }
    if (signer.toLowerCase() !== payer.toLowerCase()) {
        return invalid("signature_payer_mismatch", payer, balance, "Payment signature does not recover to payer");
    }
    return {
        isValid: true,
        payer,
        extra: { balance: balance.toString() },
    };
}
function validateAcceptedMatchesRequirements(accepted, requirements) {
    if (accepted.scheme !== "exact" || requirements.scheme !== "exact") {
        return {
            invalidReason: "unsupported_scheme",
            invalidMessage: "Unsupported scheme",
        };
    }
    if (accepted.network !== requirements.network) {
        return {
            invalidReason: "payment_network_mismatch",
            invalidMessage: `Payment accepted network mismatch: ${accepted.network} !== ${requirements.network}`,
        };
    }
    if (accepted.amount !== requirements.amount) {
        return {
            invalidReason: "payment_amount_mismatch",
            invalidMessage: `Payment accepted amount mismatch: ${accepted.amount} !== ${requirements.amount}`,
        };
    }
    if (accepted.asset.toLowerCase() !== requirements.asset.toLowerCase()) {
        return {
            invalidReason: "payment_asset_mismatch",
            invalidMessage: `Payment accepted asset mismatch: ${accepted.asset} !== ${requirements.asset}`,
        };
    }
    if (accepted.payTo.toLowerCase() !== requirements.payTo.toLowerCase()) {
        return {
            invalidReason: "payment_recipient_mismatch",
            invalidMessage: `Payment accepted payTo mismatch: ${accepted.payTo} !== ${requirements.payTo}`,
        };
    }
    if (accepted.maxTimeoutSeconds !== requirements.maxTimeoutSeconds) {
        return {
            invalidReason: "payment_requirements_mismatch",
            invalidMessage: `Payment accepted maxTimeoutSeconds mismatch: ${accepted.maxTimeoutSeconds} !== ${requirements.maxTimeoutSeconds}`,
        };
    }
    for (const key of ["name", "version", "assetTransferMethod"]) {
        const acceptedValue = accepted.extra[key];
        const requirementsValue = requirements.extra[key];
        if ((acceptedValue !== undefined || requirementsValue !== undefined) && acceptedValue !== requirementsValue) {
            return {
                invalidReason: "payment_requirements_mismatch",
                invalidMessage: `Payment accepted extra.${key} mismatch: ${String(acceptedValue)} !== ${String(requirementsValue)}`,
            };
        }
    }
    return undefined;
}
function invalid(invalidReason, payer, balance, invalidMessage) {
    return {
        isValid: false,
        invalidReason,
        ...(invalidMessage ? { invalidMessage } : {}),
        ...(payer ? { payer } : {}),
        ...(balance !== undefined ? { extra: { balance: balance.toString() } } : {}),
    };
}
function chainIdFromNetwork(network) {
    return network === "eip155:1776" ? 1776 : 1439;
}
function clientFor(network) {
    return createPublicClient({
        chain: getViemChain(network),
        transport: http(),
    });
}
async function defaultBalanceReader(asset, payer, network) {
    return getTokenBalance(clientFor(network), asset, payer);
}
async function defaultNonceReader(asset, payer, nonce, network) {
    return readNonceUsed(clientFor(network), asset, payer, nonce);
}
async function defaultTokenNameReader(asset, network) {
    return getTokenName(clientFor(network), asset);
}
function validateDomainExtra(extra, asset, network) {
    const token = getTokenByAddress(network, asset);
    const name = typeof extra?.name === "string" ? extra.name : undefined;
    const version = typeof extra?.version === "string" ? extra.version : undefined;
    const chainId = readChainIdExtra(extra);
    const verifyingContract = typeof extra?.verifyingContract === "string" ? extra.verifyingContract : undefined;
    const primaryType = typeof extra?.primaryType === "string" ? extra.primaryType : undefined;
    if (name !== undefined && token !== undefined && name !== token.name) {
        return "Payment EIP-712 domain mismatch";
    }
    if (version !== undefined && token !== undefined && version !== (token.eip712Version ?? "2")) {
        return "Payment EIP-712 domain mismatch";
    }
    if (chainId === "invalid") {
        return "Malformed chainId in payment EIP-712 domain extra";
    }
    if (chainId !== undefined && chainId !== BigInt(chainIdFromNetwork(network))) {
        return "Payment EIP-712 domain mismatch";
    }
    if (verifyingContract !== undefined && verifyingContract.toLowerCase() !== asset.toLowerCase()) {
        return "Payment EIP-712 domain mismatch";
    }
    if (primaryType !== undefined && primaryType !== "TransferWithAuthorization") {
        return "Payment EIP-712 domain mismatch";
    }
    return undefined;
}
function readVersionExtra(extra) {
    const version = extra?.version;
    return typeof version === "string" && version.length > 0 ? version : undefined;
}
function readChainIdExtra(extra) {
    const value = extra?.chainId;
    if (value === undefined) {
        return undefined;
    }
    if (typeof value === "number" && Number.isInteger(value) && value >= 0) {
        return BigInt(value);
    }
    if (typeof value === "string" && /^\d+$/.test(value)) {
        return BigInt(value);
    }
    return "invalid";
}
