// @injectivelabs/x402/dist/facilitator/index.js
/**
 * x402 Facilitator for Injective EVM.
 *
 * Exposes `verify()` and `settle()` — the two operations a facilitator server
 * must perform on behalf of resource servers.
 *
 * verify() — checks signature, balance, nonce, and time window off-chain.
 * settle() — submits transferWithAuthorization on-chain and waits for receipt.
 */
import { createPublicClient, createWalletClient, http, } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { getRpcUrl, getViemChain, isInjectiveNetwork, TOKENS, } from "../networks/index.js";
import { encodeTransferWithAuthorizationCalldata, getTokenBalance, getTokenName, isNonceUsed, } from "../eip3009.js";
import { createFacilitatorRequest, normalizeFacilitatorRequest, verifyPaymentRequest, } from "../protocol/index.js";
// ─── Facilitator class ────────────────────────────────────────────────────────
/**
 * Default minimum payment floors, keyed by asset address in smallest units.
 *
 * Operators should set `config.minPaymentPerAsset` based on their own gas and
 * pricing model. The library does not impose an asset-specific floor.
 */
const DEFAULT_MIN_PAYMENT_PER_ASSET = {};
export class InjectiveFacilitator {
    config;
    publicClients = new Map();
    walletClients = new Map();
    constructor(config) {
        this.config = {
            privateKey: config.privateKey,
            rpcUrl: config.rpcUrl ?? "",
            confirmations: config.confirmations ?? 1,
            minPaymentPerAsset: { ...DEFAULT_MIN_PAYMENT_PER_ASSET, ...(config.minPaymentPerAsset ?? {}) },
            allowedAssets: config.allowedAssets?.map((a) => a.toLowerCase()),
        };
    }
    /**
     * Compute the allowlist for `network`.
     *   - Explicit config wins, even if empty (empty = disabled).
     *   - Omitted config falls back to every EIP-3009 token in our registry.
     * Returns `undefined` only when the allowlist is explicitly disabled via
     * `allowedAssets: []`.
     */
    resolveAllowedAssets(network) {
        if (this.config.allowedAssets !== undefined) {
            return this.config.allowedAssets.length === 0
                ? undefined
                : new Set(this.config.allowedAssets);
        }
        const networkTokens = TOKENS[network] ?? {};
        const defaults = Object.values(networkTokens)
            .filter((t) => t.eip3009 && t.address !== "0x0000000000000000000000000000000000000000")
            .map((t) => t.address.toLowerCase());
        return new Set(defaults);
    }
    getPublicClient(network) {
        if (!this.publicClients.has(network)) {
            const rpcUrl = this.config.rpcUrl || getRpcUrl(network);
            this.publicClients.set(network, createPublicClient({
                chain: getViemChain(network),
                transport: http(rpcUrl),
            }));
        }
        return this.publicClients.get(network);
    }
    getWalletClient(network) {
        if (!this.walletClients.has(network)) {
            const rpcUrl = this.config.rpcUrl || getRpcUrl(network);
            const account = privateKeyToAccount(this.config.privateKey);
            this.walletClients.set(network, createWalletClient({
                account,
                chain: getViemChain(network),
                transport: http(rpcUrl),
            }));
        }
        return this.walletClients.get(network);
    }
    /**
     * Verify a payment payload before settlement.
     *
     * Checks (in order):
     *   1. Network is a supported Injective EVM network
     *   2. Scheme is "exact"
     *   3. Authorization time window is valid (not expired, not future)
     *   4. Nonce has not been used (replay protection)
     *   5. Payer has sufficient token balance
     */
    async verify(req) {
        const { paymentRequirements } = req;
        const allowedAssets = isInjectiveNetwork(paymentRequirements.network)
            ? [...(this.resolveAllowedAssets(paymentRequirements.network) ?? [])]
            : undefined;
        return verifyPaymentRequest(normalizeFacilitatorRequest(createFacilitatorRequest(req)), {
            allowedAssets,
            minPaymentPerAsset: this.config.minPaymentPerAsset,
            getBalance: async (asset, payer, network) => getTokenBalance(this.getPublicClient(network), asset, payer),
            isNonceUsed: async (asset, payer, nonce, network) => isNonceUsed(this.getPublicClient(network), asset, payer, nonce),
            getTokenName: async (asset, network) => getTokenName(this.getPublicClient(network), asset),
        });
    }
    /**
     * Settle a verified payment by submitting transferWithAuthorization on-chain.
     *
     * Supports both EOA signatures (v,r,s) and smart wallet signatures (bytes).
     */
    async settle(req) {
        const { paymentPayload, paymentRequirements } = req;
        const payload = paymentPayload.payload;
        const requirements = paymentRequirements;
        const network = requirements.network;
        // Quick pre-flight verify
        const verifyResult = await this.verify(req);
        if (!verifyResult.isValid) {
            return {
                success: false,
                errorReason: verifyResult.invalidReason,
                errorMessage: verifyResult.invalidMessage,
                payer: verifyResult.payer,
                transaction: "",
                network: req.paymentRequirements.network,
            };
        }
        const walletClient = this.getWalletClient(network);
        const publicClient = this.getPublicClient(network);
        const tokenAddress = requirements.asset;
        try {
            let txHash;
            const data = encodeTransferWithAuthorizationCalldata(payload);
            const account = walletClient.account;
            txHash = await walletClient.sendTransaction({
                chain: getViemChain(network),
                account,
                to: tokenAddress,
                data,
            });
            // Wait for on-chain confirmation
            const receipt = await publicClient.waitForTransactionReceipt({
                hash: txHash,
                confirmations: this.config.confirmations,
            });
            if (receipt.status === "reverted") {
                return {
                    success: false,
                    errorReason: "transaction_reverted",
                    errorMessage: "Transaction reverted on-chain",
                    payer: payload.authorization.from,
                    transaction: "",
                    network,
                };
            }
            return {
                success: true,
                transaction: txHash,
                network,
                payer: payload.authorization.from,
                amount: requirements.amount,
                extra: { blockNumber: receipt.blockNumber.toString() },
            };
        }
        catch (e) {
            return {
                success: false,
                errorReason: "settlement_failed",
                errorMessage: e.message,
                payer: payload.authorization.from,
                transaction: "",
                network,
            };
        }
    }
}
// ─── Standalone functions (for simple use-cases) ──────────────────────────────
/**
 * Create a facilitator bound to a single Injective network.
 */
export function createFacilitator(network, config) {
    const f = new InjectiveFacilitator(config);
    return {
        verify: (req) => f.verify(req),
        settle: (req) => f.settle(req),
    };
}
