// @injectivelabs/x402/dist/eip3009.js
/**
 * EIP-3009 and EIP-712 utilities for Injective EVM.
 *
 * Handles signing and encoding of transferWithAuthorization payloads.
 */
import { encodeFunctionData, parseSignature, recoverTypedDataAddress, } from "viem";
// ─── ABI fragments ────────────────────────────────────────────────────────────
export const EIP3009_ABI = [
    {
        name: "transferWithAuthorization",
        type: "function",
        stateMutability: "nonpayable",
        inputs: [
            { name: "from", type: "address" },
            { name: "to", type: "address" },
            { name: "value", type: "uint256" },
            { name: "validAfter", type: "uint256" },
            { name: "validBefore", type: "uint256" },
            { name: "nonce", type: "bytes32" },
            { name: "v", type: "uint8" },
            { name: "r", type: "bytes32" },
            { name: "s", type: "bytes32" },
        ],
        outputs: [],
    },
    {
        name: "transferWithAuthorization",
        type: "function",
        stateMutability: "nonpayable",
        inputs: [
            { name: "from", type: "address" },
            { name: "to", type: "address" },
            { name: "value", type: "uint256" },
            { name: "validAfter", type: "uint256" },
            { name: "validBefore", type: "uint256" },
            { name: "nonce", type: "bytes32" },
            { name: "signature", type: "bytes" },
        ],
        outputs: [],
    },
    {
        name: "authorizationState",
        type: "function",
        stateMutability: "view",
        inputs: [
            { name: "authorizer", type: "address" },
            { name: "nonce", type: "bytes32" },
        ],
        outputs: [{ name: "", type: "bool" }],
    },
    {
        name: "balanceOf",
        type: "function",
        stateMutability: "view",
        inputs: [{ name: "account", type: "address" }],
        outputs: [{ name: "", type: "uint256" }],
    },
    {
        name: "name",
        type: "function",
        stateMutability: "view",
        inputs: [],
        outputs: [{ name: "", type: "string" }],
    },
];
// ─── EIP-712 typed data ───────────────────────────────────────────────────────
export const TRANSFER_WITH_AUTHORIZATION_TYPE = {
    TransferWithAuthorization: [
        { name: "from", type: "address" },
        { name: "to", type: "address" },
        { name: "value", type: "uint256" },
        { name: "validAfter", type: "uint256" },
        { name: "validBefore", type: "uint256" },
        { name: "nonce", type: "bytes32" },
    ],
};
export function buildTransferWithAuthorizationTypedData(input) {
    return {
        domain: {
            name: input.tokenName,
            version: input.tokenVersion ?? "2",
            chainId: input.chainId,
            verifyingContract: input.tokenAddress,
        },
        types: TRANSFER_WITH_AUTHORIZATION_TYPE,
        primaryType: "TransferWithAuthorization",
        message: {
            from: input.authorization.from,
            to: input.authorization.to,
            value: BigInt(input.authorization.value),
            validAfter: BigInt(input.authorization.validAfter),
            validBefore: BigInt(input.authorization.validBefore),
            nonce: input.authorization.nonce,
        },
    };
}
export function encodeTransferWithAuthorizationCalldata(input) {
    const auth = input.authorization;
    const signature = input.signature;
    if (signature.length === 132) {
        const { v, r, s } = parseSignature(signature);
        return encodeFunctionData({
            abi: EIP3009_ABI,
            functionName: "transferWithAuthorization",
            args: [
                auth.from,
                auth.to,
                BigInt(auth.value),
                BigInt(auth.validAfter),
                BigInt(auth.validBefore),
                auth.nonce,
                Number(v),
                r,
                s,
            ],
        });
    }
    return encodeFunctionData({
        abi: EIP3009_ABI,
        functionName: "transferWithAuthorization",
        args: [
            auth.from,
            auth.to,
            BigInt(auth.value),
            BigInt(auth.validAfter),
            BigInt(auth.validBefore),
            auth.nonce,
            signature,
        ],
    });
}
// ─── Nonce generation ─────────────────────────────────────────────────────────
/** Generate a cryptographically random 32-byte nonce. */
export function createNonce() {
    const bytes = new Uint8Array(32);
    crypto.getRandomValues(bytes);
    return `0x${Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("")}`;
}
// ─── Signing ──────────────────────────────────────────────────────────────────
/**
 * Sign an EIP-3009 transferWithAuthorization using EIP-712.
 *
 * @param walletClient  viem WalletClient with a connected account
 * @param tokenAddress  ERC-3009 token contract
 * @param tokenName     Token name (for domain separator, e.g. "USD Coin")
 * @param chainId       Numeric chain ID (1776 mainnet, 1439 testnet)
 * @param auth          The authorization parameters
 * @returns             EIP-712 signature as hex
 */
export async function signAuthorization(walletClient, tokenAddress, tokenName, chainId, auth, version = "2") {
    const account = walletClient.account;
    if (!account)
        throw new Error("WalletClient has no account");
    return walletClient.signTypedData({
        account,
        domain: {
            name: tokenName,
            version,
            chainId,
            verifyingContract: tokenAddress,
        },
        types: TRANSFER_WITH_AUTHORIZATION_TYPE,
        primaryType: "TransferWithAuthorization",
        message: {
            from: auth.from,
            to: auth.to,
            value: auth.value,
            validAfter: auth.validAfter,
            validBefore: auth.validBefore,
            nonce: auth.nonce,
        },
    });
}
// ─── Signature verification ───────────────────────────────────────────────────
/**
 * Recover the signer address from an EIP-3009 transferWithAuthorization signature.
 *
 * @param tokenAddress   ERC-3009 token contract
 * @param tokenName      Token name used in the EIP-712 domain (e.g. "USDC")
 * @param chainId        Numeric chain ID (1776 mainnet, 1439 testnet)
 * @param auth           The authorization parameters
 * @param signature      EIP-712 hex signature from the payer
 * @param version        EIP-712 domain version (default "2", used by Circle FiatTokenV2_2)
 * @returns              Recovered signer address
 */
export async function verifyAuthorization(tokenAddress, tokenName, chainId, auth, signature, version = "2") {
    return recoverTypedDataAddress({
        domain: {
            name: tokenName,
            version,
            chainId,
            verifyingContract: tokenAddress,
        },
        types: TRANSFER_WITH_AUTHORIZATION_TYPE,
        primaryType: "TransferWithAuthorization",
        message: {
            from: auth.from,
            to: auth.to,
            value: auth.value,
            validAfter: auth.validAfter,
            validBefore: auth.validBefore,
            nonce: auth.nonce,
        },
        signature,
    });
}
// ─── On-chain state reads ─────────────────────────────────────────────────────
export async function getTokenBalance(client, tokenAddress, owner) {
    return client.readContract({
        address: tokenAddress,
        abi: EIP3009_ABI,
        functionName: "balanceOf",
        args: [owner],
    });
}
export async function isNonceUsed(client, tokenAddress, authorizer, nonce) {
    return client.readContract({
        address: tokenAddress,
        abi: EIP3009_ABI,
        functionName: "authorizationState",
        args: [authorizer, nonce],
    });
}
export async function getTokenName(client, tokenAddress) {
    return client.readContract({
        address: tokenAddress,
        abi: EIP3009_ABI,
        functionName: "name",
        args: [],
    });
}
