// @injectivelabs/x402/dist/index.js
/**
 * @injectivelabs/x402 — x402 Payment Protocol for Injective EVM
 *
 * @example Server (Express)
 * ```ts
 * import express from "express";
 * import { injectivePaymentMiddleware } from "@injectivelabs/x402/middleware";
 *
 * const app = express();
 * app.use(injectivePaymentMiddleware({
 *   "GET /api/data": {
 *     description: "Premium data feed",
 *     accepts: [{
 *       network: "eip155:1776",
 *       asset: "0xUSCDAddress",
 *       amount: "1000000", // 1 USDC
 *     }],
 *   },
 * }, { facilitator: { privateKey: process.env.PRIVATE_KEY } }));
 * ```
 *
 * @example Client
 * ```ts
 * import { createInjectiveClient } from "@injectivelabs/x402/client";
 *
 * const client = createInjectiveClient({ privateKey: process.env.PRIVATE_KEY });
 * const response = await client.fetch("https://api.example.com/api/data");
 * ```
 */
// Networks
export { injectiveEvm, injectiveEvmTestnet, INJECTIVE_MAINNET_CAIP2, INJECTIVE_TESTNET_CAIP2, TOKENS, isInjectiveNetwork, getViemChain, getRpcUrl, getToken, getTokenByAddress, registerToken, } from "./networks/index.js";
// Schemas (zod) for runtime validation at trust boundaries
export { Eip3009AuthorizationSchema, InjectivePaymentPayloadSchema, ResourceInfoSchema, PaymentRequirementsSchema, PaymentRequiredSchema, PaymentPayloadSchema, FacilitatorVerifyRequestSchema, FacilitatorSettleRequestSchema, VerifyRequestSchema, SettleRequestSchema, VerifyResponseSchema, SettleResponseSchema, SupportedKindSchema, SupportedResponseSchema, MAX_PAYLOAD_BYTES, assertPayloadSize, } from "./schemas.js";
// EIP-3009 utilities
export { EIP3009_ABI, TRANSFER_WITH_AUTHORIZATION_TYPE, buildTransferWithAuthorizationTypedData, encodeTransferWithAuthorizationCalldata, createNonce, signAuthorization, verifyAuthorization, getTokenBalance, isNonceUsed, getTokenName, } from "./eip3009.js";
export { createFacilitatorRequest, createPaymentPayload, normalizeFacilitatorRequest, toInnerVerifyRequest, verifyPaymentRequest, } from "./protocol/index.js";
// Facilitator
export { InjectiveFacilitator, createFacilitator, } from "./facilitator/index.js";
// Client
export { createPayment, createInjectiveClient, encodePaymentSignatureHeader, decodePaymentSignatureHeader, parsePaymentRequired, parsePaymentResponseHeader, } from "./client/index.js";
// Middleware
export { injectivePaymentMiddleware, } from "./middleware/index.js";
