// @injectivelabs/x402/dist/schemas.js
/**
 * Zod schemas for validating payloads that cross a trust boundary.
 *
 * Any string coming in over HTTP (base64 `PAYMENT-SIGNATURE`, JSON bodies on
 * /verify and /settle, 402 responses from third-party servers) should be
 * parsed with these schemas before we believe it. This catches malformed
 * input early, gives callers actionable error messages, and caps the
 * incidental DoS surface from oversized or malformed bodies.
 */
import { z } from "zod";
/** Narrow zod outputs to viem's template-literal types so downstream
 *  consumers get the right types after `.parse()`. */
const Hex = z.string().regex(/^0x[0-9a-fA-F]+$/, "expected 0x-prefixed hex").transform((v) => v);
const Address = z.string().regex(/^0x[0-9a-fA-F]{40}$/, "expected 20-byte address").transform((v) => v);
const Bytes32 = z.string().regex(/^0x[0-9a-fA-F]{64}$/, "expected 32-byte hex").transform((v) => v);
const DecimalU = z.string().regex(/^\d+$/, "expected non-negative integer string");
const CAIP2Eth = z.string().regex(/^eip155:\d+$/, "expected CAIP-2 eip155:<chainId>");
const UnknownRecord = z.record(z.unknown());
const OptionalRecord = UnknownRecord.nullish().transform((v) => v ?? undefined);
const DefaultRecord = UnknownRecord.nullish().transform((v) => v ?? {});
export const Eip3009AuthorizationSchema = z.object({
    from: Address,
    to: Address,
    value: DecimalU,
    validAfter: DecimalU,
    validBefore: DecimalU,
    nonce: Bytes32,
});
export const InjectivePaymentPayloadSchema = z.object({
    signature: Hex,
    authorization: Eip3009AuthorizationSchema,
});
export const ResourceInfoSchema = z.object({
    url: z.string().min(1),
    description: z.string().optional(),
    mimeType: z.string().optional(),
    serviceName: z.string().min(1).max(32).regex(/^[\x20-\x7e]+$/).optional(),
    tags: z.array(z.string().min(1).max(32).regex(/^[\x20-\x7e]+$/)).max(5).optional(),
    iconUrl: z.string().max(2048).optional(),
});
export const PaymentRequirementsSchema = z.object({
    scheme: z.literal("exact"),
    network: CAIP2Eth,
    amount: DecimalU,
    asset: Address,
    payTo: Address,
    maxTimeoutSeconds: z.number().positive(),
    extra: DefaultRecord,
});
export const PaymentRequiredSchema = z.object({
    x402Version: z.literal(2),
    error: z.string().optional(),
    resource: ResourceInfoSchema,
    accepts: z.array(PaymentRequirementsSchema).min(1),
    extensions: OptionalRecord,
});
export const PaymentPayloadSchema = z.object({
    x402Version: z.literal(2),
    resource: ResourceInfoSchema.optional(),
    accepted: PaymentRequirementsSchema,
    payload: InjectivePaymentPayloadSchema,
    extensions: OptionalRecord,
});
export const FacilitatorVerifyRequestSchema = z.object({
    x402Version: z.literal(2),
    paymentPayload: PaymentPayloadSchema,
    paymentRequirements: PaymentRequirementsSchema,
});
export const FacilitatorSettleRequestSchema = FacilitatorVerifyRequestSchema;
export const VerifyRequestSchema = z.object({
    paymentPayload: PaymentPayloadSchema,
    paymentRequirements: PaymentRequirementsSchema,
});
export const SettleRequestSchema = VerifyRequestSchema;
export const VerifyResponseSchema = z.object({
    isValid: z.boolean(),
    invalidReason: z.string().optional(),
    invalidMessage: z.string().optional(),
    payer: Address.optional(),
    extensions: OptionalRecord,
    extra: OptionalRecord,
});
export const SettleResponseSchema = z.object({
    success: z.boolean(),
    errorReason: z.string().optional(),
    errorMessage: z.string().optional(),
    payer: Address.optional(),
    transaction: z.union([Hex, z.literal("")]),
    network: CAIP2Eth,
    amount: DecimalU.optional(),
    extensions: OptionalRecord,
    extra: OptionalRecord,
});
export const SupportedKindSchema = z.object({
    x402Version: z.literal(2),
    scheme: z.literal("exact"),
    network: CAIP2Eth,
    extra: OptionalRecord,
});
export const SupportedResponseSchema = z.object({
    kinds: z.array(SupportedKindSchema),
    extensions: z.array(z.string()),
    signers: z.record(z.array(Address)),
});
/**
 * Guard against pathological inputs on the HTTP boundary (facilitator
 * endpoints, base64-decoded headers). 8 KB is generous for a protocol
 * payload — the canonical shape is a few hundred bytes.
 */
export const MAX_PAYLOAD_BYTES = 8 * 1024;
export function assertPayloadSize(raw) {
    if (raw.length > MAX_PAYLOAD_BYTES) {
        throw new Error(`x402 payload exceeds ${MAX_PAYLOAD_BYTES} bytes (got ${raw.length}); refusing to parse.`);
    }
}
